crypto/src/signature.rs

```diff
@@ -0,0 +1,85 @@
+// SPDX-FileCopyrightText: 2023 TriliTech <contact@trili.tech>
+// Copyright (c) SimpleStaking, Viable Systems and Tezedge Contributors
+//
+// Ported from octez: lib_crypto/signature_v1.ml
+//
+// Copyright (c) 2018 Dynamic Ledger Solutions, Inc. <contact@tezos.com>
+// Copyright (c) 2020 Metastate AG <hello@metastate.dev>
+// Copyright (c) 2022 Nomadic Labs <contact@nomadic-labs.com>
+//
+// SPDX-License-Identifier: MIT
+
+use crate::base58::FromBase58CheckError;
+use crate::hash::{
+    BlsSignature, Ed25519Signature, FromBytesError, HashTrait, HashType, P256Signature,
+    Secp256k1Signature, UnknownSignature,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Signature {
```

Comment:
This type is seemingly missing a few instances I would expect:

- `FromStr`
- `HashTrait` (possibly, perhaps doesn't quite fit, but would be useful to have)
- `Into<Hash>` (via `impl From<Signature> for Hash`)
- `TryFrom<&[u8]>`
- `TryFrom<Hash>`
- `TryFrom<&str>`

Reply:
`HashTrait` can't be implemented, since we don't statically know the `HashType`, but I can return an `impl HashTrait`

Reply:
Have implemented the rest though

Reply:
Ah, right, we had this issue in MIR, too. Seems a bit unfortunate we don't have to/from base58 conversions via a trait, but amending that is definitely out of scope here.